server/accounts.go

```diff
@@ -1344,26 +1345,85 @@ func (a *Account) AddServiceImportWithClaim(destination *Account, from, to strin
 		return ErrServiceImportAuthorization
 	}
 
-	if a.importFormsCycle(destination, from, to) {
-		return ErrServiceImportFormsCycle
+	// Check if this introduces a cycle before proceeding.
+	if a.serviceImportFormsCycle(destination, from) {
+		return ErrImportFormsCycle
 	}
 
 	_, err := a.addServiceImport(destination, from, to, imClaim)
 	return err
 }
 
-// Detects if we have a cycle.
-func (a *Account) importFormsCycle(destination *Account, from, to string) bool {
-	// Check that what we are importing is not something we also export.
-	if a.serviceExportOverlaps(to) {
-		// So at this point if destination account is also importing from us, that forms a cycle.
-		if destination.serviceImportOverlaps(from) {
+func (a *Account) serviceImportFormsCycle(dest *Account, from string) bool {
+	return dest.checkServiceImportsForCycles(from, map[string]bool{a.Name: true})
+}
+
+func (a *Account) checkServiceImportsForCycles(from string, visited map[string]bool) bool {
+	a.mu.RLock()
+	for _, si := range a.imports.services {
+		if SubjectsCollide(from, si.to) {
+			a.mu.RUnlock()
+			if visited[si.acc.Name] {
+				return true
+			}
+			// Push ourselves and check si.acc
+			visited[a.Name] = true
+			if subjectIsSubsetMatch(si.from, from) {
+				from = si.from
+			}
+			if si.acc.checkServiceImportsForCycles(from, visited) {
+				return true
+			}
+			a.mu.RLock()
+		}
+	}
+	a.mu.RUnlock()
+	return false
+}
+
+func (a *Account) streamImportFormsCycle(dest *Account, to string) bool {
+	return dest.checkStreamImportsForCycles(to, map[string]bool{a.Name: true})
+}
+
+// Lock should be held.
+func (a *Account) hasStreamExportMatching(to string) bool {
+	for subj := range a.exports.streams {
+		if subjectIsSubsetMatch(to, subj) {
 			return true
 		}
 	}
 	return false
 }
 
+func (a *Account) checkStreamImportsForCycles(to string, visited map[string]bool) bool {
```

Comment:
So there is no communality between the two that could be added or used to avoid duplicating what is essentially the same code?

Reply:
I had them combined for awhile but was hard to read back TBH, so split them back up again. The way we track down the cycles is different enough IMO.

Reply:
duplicated code is never good

Reply:
Read them a but more closely..

Reply:
So why is it that streams are being iterated over in a separate pass and services are not?

Reply:
Not following that statement, we iterate over each one looking for cycles separately.